Initial implementation

cottsay opened the pull request.

cottsay committed:
Add mechanism to read pyproject.toml specs

cottsay committed:
Implement PEP 517 hook caller

cottsay committed:
Implement PEP 621 identification and augmentation

cottsay committed:
Implement decoration extension for enhancing hook calls

cottsay committed:
Implement basic mechanism for installing Python wheels

cottsay committed:
Implement mechanism for reading python METADATA files

cottsay committed:
Implement PEP 517 identification and augmentation

cottsay committed:
Implement PEP 517 build task

cottsay committed:
Implement test task for python.project packages

cottsay committed:
Update README.rst to reflect current status

cottsay committed:
Add support for in-tree build backends

cottsay committed:
Add support for setuptools script directory override

cottsay committed:
Fix installation of data files from wheels

cottsay committed:
More robust wheel directory creation

cottsay committed:
Add a PEP 517 identification fallback for legacy setuptools

cottsay committed:
Implement RECORD file installation for wheels

cottsay committed:
Add missing test dependency on pytest-asyncio

cottsay committed:
Forward subprocess stderr in hook tests

cottsay committed:
Forward handles instead of fds on Windows

cottsay committed:
Add hooks and scripts to RECORD file

cottsay committed:
Add 'INSTALLER' file to dist-info

cottsay committed:
Bump colcon-core dependency version

cottsay committed:
Cache metadata using getter

This approach is already used by the existing Python metadata caching to
reduce the descriptor size.

cottsay committed:
Return dist-info directory path from wheel installation

We can easily determine the RECORD path from the dist-info directory, so
returning the directory specifically is more straightforward.

cottsay committed:
Implement hook listing functionality

cottsay committed:
Implement PEP 660 editable installs

cottsay committed:
Add usage instructions

cottsay committed:
Invoke hook subprocess as script instead of module

Module-based invocation works well when the subprocess interpreter has
the same Python PATH and modules as the main process, but this isn't
always the case (i.e. pytest).

It's safer to import the module we want to call in the main process and
directly invoke that script in the subprocess.

One behavioral difference between the module and script is that scripts
resolve import starting with the script's directory. To avoid
collisions, this change moves the scripts into a new subdirectory.

cottsay committed:
Add more tests, make the prototype easier to use

cottsay commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=18862)](http://ci.ros2.org/job/ci_linux/18862/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=13400)](http://ci.ros2.org/job/ci_linux-aarch64/13400/)
* Linux-rhel8 [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=426)](https://ci.ros2.org/job/ci_linux-rhel/426/)
* Linux-rhel9 [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=423)](https://ci.ros2.org/job/ci_linux-rhel/423/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=19615)](http://ci.ros2.org/job/ci_windows/19615/)
* Windows-debug [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_windows&build=19616)](https://ci.ros2.org/job/ci_windows/19616/)

cottsay committed:
Suppress an upstream deprecation

cottsay committed:
Close event loops to avoid ResourceWarning

cottsay committed:
Set the thread's event loop

cottsay committed:
Update test requirements

cottsay committed:
Python 3.6 compatibility

cottsay committed:
Let the Python 3.6 tests fail for now

cottsay committed:
Use colcon's new_event_loop for async benchmarking

cottsay committed:
Deal with older setuptools versions

cottsay committed:
Drop dependency on pytest-asyncio

cottsay committed:
Revert "Let the Python 3.6 tests fail for now"

This reverts commit 21e7d1dc1bf3edfa89c0b4cf776a37fb9adc0959.

cottsay committed:
How did that get there?

cottsay committed:
Add test for build task

cottsay committed:
Use importlib.metadata to find and uninstall packages

cottsay committed:
Specifically set thread's event loop prior to benchmarking

cottsay committed:
Significantly more robust package removal

cottsay committed:
Add a slightly hacky way to remove egg-info

cottsay committed:
Add (new) usage instructions

cottsay committed:
...make that a code block

cottsay committed:
Fixes for Windows

cottsay committed:
fixup!

cottsay commented:
Thanks for giving this a shot, @emanuelbuholzer.

That error happens with the version of `flake8` is too recent. I believe that the changes showed up in 3.8.0, so if you downgrade to a version older than that, you should see improvement.

emanuelbuholzer commented:
Thanks @cottsay for the feedback and your work on this. I deleted my original comment, because the tests ran all fine within my CI pipeline and I thought I was maybe a bit too eager. I'm creating an example workspace with some example packages to test some use cases with it further. Hopefully I'll have some more feedback soon.

emanuelbuholzer commented on colcon_python_project/task/ament_python/project/build.py:
```diff
@@ -0,0 +1,108 @@
+
+        # check if the package index and manifest are being installed
+        data_files = get_data_files_mapping(
+            setup_py_data.get('data_files', []) or [])
```
I wonder if we need an alternative approach of getting the package index and manifest, or otherwise not warn about the fallback and it being removed in the future. 

If you migrate away from the legacy `setup.py` build you most likely don't want to have a `setup.py` in your package. My testing showed that this is not possible. From what I understand, this is to keep compatibility with legacy builds or versions of tools that don’t support certain packaging standards (see [setuptools documentation](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#setupcfg-caveats)). If this is the case, I think it'd be great to follow the documentation and to try to keep it a simple `setup()` call.

With this current implementation however the package index and manifest can only be read if you have a setup.py file with its `data_files=[('share/ament_index/resource_index/packages', ['resource/' + package_name]), ('share/' + package_name, ['package.xml'])]` mapping. Otherwise you'll be prompted with the warnings about the fallback and it being removed in the future. I'm unsure what's the best way to solve this and would love to hear your thoughts about this. From what I understand, the traditional data files support is deprecated (see [setuptools documentation](https://setuptools.pypa.io/en/latest/userguide/datafiles.html)).

cottsay committed:
Fix rebuild of packages without top_level.txt

cottsay committed:
Fake setuptools options so that test verb works

Partially because colcon dependencies don't carry context around what
package type the dependency originated from, it's difficult for the
python testing step to reliably determine the test dependencies using
abstract mechanisms.

For the time being, the easiest thing to do to enable those plugins for
packages discovered by standards-based extensions is to mimic the
setuptools metadata format for dependencies.

cottsay committed:
Drop partial hook methods, rally around 'call_hook'

Having multiple paths to calling a specific hook has immediately
backfired after a previous change started using 'call_hook' instead of
the individual partial method, which routed around the setuptools
decorator. It was a nice idea, but isn't really needed.

cottsay committed:
Quiet pydocstyle warnings during flake8 testing

This extension seems to modify the global logging config more than other
extensions do, and the result is very verbose output from the flake8
tests. Suppressing pydocstyle debug messages seems to maket things
reasonable again.

cottsay committed:
Address a new deprecation in colcon-core

cottsay committed:
Fix spelling test

cottsay committed:
Suppress pkg_resources deprecation in flake8_import_order

cottsay committed:
Fix priority on ament package augmentation

High number == run sooner, low number == run later

cottsay committed:
Fix setuptools escape hatch for earlier than v64

Also add the unfortunate note about symlink installs and data_files.

cottsay committed:
Suppress an irrelevant informational message from setuptools

cottsay committed:
Handle hook call without explicit environment vars

Fixes: d48d6832dcbf4

david-dorf commented:
Interested in testing/reviewing this. Is the intention that I would follow the instructions here: https://github.com/colcon/colcon-python-project/tree/devel , create a ROS 2 package, delete the setup.py and replace it with pyproject.toml, and see if it will build properly?

Briancbn commented:
Hi @cottsay, thanks for this project. Unfortunately, it seems that [this thread is closed](https://discourse.openrobotics.org/t/call-for-testing-standards-based-python-packaging-with-colcon/32008). We have been using this project for a while now, here are some of the issues and suggestion
- Since this repo disables `python` as a package type, it invalidated some of our previous "hack" to have a `colcon.pkg` file to force colcon to recognize our package as a pure python package, while still getting the benefit of the `rosdep` and `package.xml` ([example](https://github.com/ros-industrial/rmf_scheduler/tree/main/rmf2_scheduler_server_py)). I wonder if it is better to not update the build method for the old `setup.py` and `setup.cfg` only package. Only update if there is a pyproject.toml specified. (no `pyproject,toml` = old behavior, has `pyproject.toml` = new behavior).
- This repo also relies on `pip` to install the build hooks such as `uv_build`, otherwise it will result in an error. Is it possible to define a different behavior for like a `uv` package that uses the `uv` executable to build using some of the hook mechanism in this project? Currently, I did [this](https://github.com/Briancbn/colcon-python-project-uv) to overwrite your build using UV. Please ignore the venv part.
- What are your thoughts on leveraging the `uv` build as a build front-end instead? This can reduce the need of installation various build hooks library from pip. It also support poetry, sckit-build-core, setuptools and more. It is also used quite extensively in the [pydev handbook](https://pydevtools.com/handbook/explanation/what-is-a-build-frontend/).
- UV also has a workspace feature similar to cargo, I implemented in [my own personal project](https://github.com/Briancbn/colcon-python-project-uv/blob/main/colcon_python_project_uv/package_identification/uv_workspace.py) similar to the cargo project. Do you think we should include in this project?
- We have encountered some issue with symlink install, but haven't got the time to take a deeper look whether is caused by this project or by uv_build.
- We have not tried a `ament_python` setup with `pyproject.toml` with this repo.

Same as @david-dorf, please let me know your plan for this project. I can help out such as writing tests and testing with different package configurations.

cottsay committed:
Update target Ubuntu/Debian suites

cottsay committed:
Add a top-level .gitignore

cottsay committed:
Add markers to linter tests, move linter imports

Test markers can be used to easily (de-)select tests, and colcon exposes
mechanisms to do so. Linters are a category of tests that are commonly
called out.

Additionally, if we move the imports for some of our single-purpose
tests into the test function, we can avoid installing the linter
dependencies entirely. This is a common case in platform packaging, where
linter errors are not actionable and the dependencies are not typically
installed.

cottsay committed:
Switch to logger.warning instead of deprecated logger.warn

cottsay committed:
Switch to SPDX license identifier in setup.cfg

cottsay committed:
Add '+upstream' suffix to published deb version

We continue to see interference between the deb packages we publish from
the colcon project and efforts to package colcon as part of mainline
Debian and Ubuntu. Using a high Debian-Version value mitigated the
problems in most cases, but was not sufficient to eliminate all of the
conflicts we're currently experiencing.

Using a debian version suffix which falls late alphabetically appears to
give our packages preference by apt. If a user enables a repository
which distributes packages created by the colcon project, it is likely
that they wish to use these packages instead of the ones packaged by
their platform.

cottsay committed:
Stop treating warnings as errors in tests

While this may have been well-intended, years of builds have
demonstrated that we only really see deprecation warnings in our
dependencies and rarely catch anything in colcon packages.

We may elect to re-enable this flag in our CI builds, but having it
enabled in the package itself only makes it more difficult to maintain
colcon packages downstream.
